### ISSUE TITLE
Unable to load custom model when using tf.keras.callbacks.ModelCheckpoint

### ISSUE THREAD

**Issue by esdrascosta** [tensorflow/tensorflow#41340]
**System information**

== check python ===================================================
python version: 3.6.9
python branch: 
python build version: ('default', 'Apr 18 2020 01:56:04')
python compiler version: GCC 8.4.0
python implementation: CPython

== check os platform ===============================================
os: Linux
os kernel version: 1 SMP Wed Feb 19 05:26:34 PST 2020
os release version: 4.19.104+
os platform: Linux-4.19.104+-x86_64-with-Ubuntu-18.04-bionic
linux distribution: ('Ubuntu', '18.04', 'bionic')
linux os distribution: ('Ubuntu', '18.04', 'bionic')
mac version: ('', ('', '', ''), '')
uname: uname_result(system='Linux', node='6fd50df0cf80', release='4.19.104+', version='#1 SMP Wed Feb 19 05:26:34 PST 2020', machine='x86_64', processor='x86_64')
architecture: ('64bit', '')
machine: x86_64

== are we in docker =============================================
Yes

== compiler =====================================================
c++ (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

== check pips ===================================================
numpy                    1.18.5         
protobuf                 3.10.0         
tensorflow               2.2.0          
tensorflow-addons        0.8.3          
tensorflow-datasets      2.1.0          
tensorflow-estimator     2.2.0          
tensorflow-gcs-config    2.2.0          
tensorflow-hub           0.8.0          
tensorflow-metadata      0.22.2         
tensorflow-privacy       0.2.2          
tensorflow-probability   0.10.0         

== check for virtualenv =========================================
False

== tensorflow import ============================================
tf.version.VERSION = 2.2.0
tf.version.GIT_VERSION = v2.2.0-0-g2b96f3662b
tf.version.COMPILER_VERSION = 7.4.0

== env ==========================================================
LD_LIBRARY_PATH /usr/lib64-nvidia
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
Mon Jul 13 14:50:08 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 450.36.06    Driver Version: 418.67       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla K80           Off  | 00000000:00:04.0 Off |                    0 |
| N/A   44C    P0    59W / 149W |    157MiB / 11441MiB |      0%      Default |
|                               |                      |                 ERR! |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+

== cuda libs  ===================================================
/usr/local/lib/python2.7/dist-packages/torch/lib/libcudart-1b201d85.so.10.1
/usr/local/lib/python3.6/dist-packages/torch/lib/libcudart-1b201d85.so.10.1
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudart_static.a
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudart.so.10.0.130
/usr/local/cuda-10.0/doc/man/man7/libcudart.7
/usr/local/cuda-10.0/doc/man/man7/libcudart.so.7
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudart.so.10.1.243
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudart_static.a
/usr/local/cuda-10.1/doc/man/man7/libcudart.7
/usr/local/cuda-10.1/doc/man/man7/libcudart.so.7

== tensorflow installed from info ==================
Name: tensorflow
Version: 2.2.0
Summary: TensorFlow is an open source machine learning framework for everyone.
Home-page: https://www.tensorflow.org/
Author-email: packages@tensorflow.org
License: Apache 2.0
Location: /usr/local/lib/python3.6/dist-packages
Required-by: fancyimpute

== python version  ==============================================
(major, minor, micro, releaselevel, serial)
(3, 6, 9, 'final', 0)

- All other system information is contained in a standalone colab notebook as cited section below

**Describe the current behavior**

When training a sub-classed keras model with tf.keras.callbacks.ModelCheckpoint, It is not possible load the saved model. 

> OSError                                   Traceback (most recent call last)
> <ipython-input-7-bd53ee47e8b4> in <module>()
> ----> 1 tf.keras.models.load_model('checkpoints/classifier')>
> 
> 1 frames
> /usr/local/lib/python3.6/dist-packages/tensorflow/python/saved_model/loader_impl.py in parse_saved_model(export_dir)
>     111                   (export_dir,
>     112                    constants.SAVED_MODEL_FILENAME_PBTXT,
> --> 113                    constants.SAVED_MODEL_FILENAME_PB))
>     114 
>     115 
> OSError: SavedModel file does not exist at: checkpoints/classifier/{saved_model.pbtxt|saved_model.pb}

 
**Describe the expected behavior**
It should be able to use the saved model for inference/deployment or other subsequent tasks

**Standalone code to reproduce the issue**
https://colab.research.google.com/drive/1mJnwoTCJOcZKiPxfY5nDfFk14S8r5IOi

**Other info / logs** Include any logs or source code that would be helpful to
 I think there are other issues related like https://github.com/tensorflow/tensorflow/issues/31057

**Comment by ravikyram** [comment]
@esdrascosta 

I see you have not saved the model before you load the model.After saving the model i am not seeing any issue.Can you please find the gist [here ](https://colab.research.google.com/gist/ravikyram/3a0e9179b21a80d4de26c7f92c55430c/untitled117.ipynb). Thanks!

**Comment by esdrascosta** [comment]
Hi @ravikyram, thanks for your feedback. The problem is when using ModelCheckpoint, the model is not saved properly. I'm concerning for long time training and saving only the best model. After saving the model once compiled, as you demonstrated on the gist calling explicit `mode.save`, indeed the `.pb` file is generated, but there is not guarantee that the model saved is "best" model saved, given the `monitor='val_accuracy'`,  after long training the last state of model might not be the best one. My guess is in the tf.keras.callbacks.ModelCheckpoint its self.

**Comment by esdrascosta** [comment]
I think may be related with this issue https://github.com/tensorflow/tensorflow/issues/37839

**Comment by gowthamkpr** [comment]
@esdrascosta Did you try using tensorflow 2.3.0?

**Comment by esdrascosta** [comment]
Hi @gowthamkpr, yes I do with the latest version `2.3.0rc1`, you can find the gist [here](https://colab.research.google.com/drive/1XuGXzPlPWvhJO8c0-sUm4UFoSImQqDzA). Thanks in advance.

**Comment by esdrascosta** [comment]
It seems something about the overwritten method `set_model`  in tf.keras.callbacks.ModelCheckpoint. Under the hood, whenever a model is set it checks: 

``` 
def set_model(self, model):
    self.model = model
    if (not self.save_weights_only and
        not model._is_graph_network and
        model.__class__.__name__ != 'Sequential'):
      self.save_weights_only = True
```  

So, by default `save_weights_only = False` and when the model is set and it satisfies that condition it changes this default behavior. I could open a PR about this, but I'm not familiar with this internal source code. If you could give me a explanation why this condition is placed there, I could contribute.

**Comment by sedghi** [comment]
Can we get an update on this?

**Comment by sushreebarsa** [comment]
Was able to replicate this issue in TF v2.5,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/52cbb9d167e6d5804e0c53c2d771c7fb/untitled133.ipynb)..Thanks !

**Comment by mohantym** [comment]
Hi @esdrascosta ! I was able to load the model by putting full path of checkpoints folder. Attaching [gist ](https://colab.sandbox.google.com/gist/mohantym/b7826839a1ab6c95f81f8ddcee58b83e/untitled133.ipynb#scrollTo=h65EkFu9RByu)in 2.7 for reference. Thanks!

**Comment by esdrascosta** [comment]
Thanks, @mohantym  I've just tested the issue and it's seems to be fixed

The version I was testing at that time was `tensorflow==2.2.0`. 
So, now this issue should have be fixed in 2.7 version.

I also tested with relative paths and worked as well. 
`
tf.keras.models.load_model(os.path.join('checkpoints', 'classifier'))
`
Thanks again.
